README.md
# Kubernetes-Openshift
General repo for Kubernetes and Openshift projects

Kubernetes and Openshift related documentation :

## 1. [Openshift Cheat Sheet](documentation/01-openshift-cheat-sheet.md)

## 2. [K8s and OCP commn issues](documentation/02-ocp-k8s-common-issues.md)

## 3. [Kubernetes Cheat Sheet](documentation/03-kubernetes-cheat-sheet.md)

Kubernetes and Openshift related projects :

## 1. [Vault](vault/README.md)

## 2. [Velero and Minio](velero-minio/README.md)

## 3. [IBM private worker](toolchain-private-worker/README.md)


toolchain-private-worker/CHANGELOG.md
# Changelog

## Remember:

### REV M.m.r, where:

- M: Mayor — Refactoring, Adding extra capabilities. This is not a fix.
- m: Minor — This is a fix. Correcting an existing functionality.
- r: Release — Programmer error. Bug.
- a/b: alpha or beta status for early access release

## Actual versions

- v2.0.0
- v1.0.0
- v1.0.0b

## [v2.0.0]
#### 04/07/2020
### Code revision
  *  Improved code
  *  Environment variables mode
  *  Dynamic environment reference

## [v1.0.0]
#### 03/07/2020
### All New Features
  *  Initial Release

## [v1.0.0b] - Early Access
#### 01/07/2020
### All New Features (only few features are working correctly)
  *  Pre-release
  *  Incomplete working state

velero-minio/resources/minio/scripts/configure_minio_server.sh
#!/bin/bash
# auto provisioning script used after minio client start-up to add minio servers and provision bucket

########## VARIABLES ##########
buckets=( vault bpm awx )
###############################

########## CODE ##########
echo "CHECK > waiting for minio port opening..."
while [ $(nc -z minio1 9000; echo $?) != "0" ]
do
  echo "IDLE > waiting a bit more..."
  sleep 5
done

echo "SUCCESS > minio1 port opened"
sleep 10

echo "PROCESS > addind minio instances to minio client"
mc config host add minio http://${MINIO_IP}:9001 ${MINIO_ACCESS_KEY} ${MINIO_SECRET_KEY}
echo "SUCCESS > minio instances added"

echo "DISPLAY > minio servers status"
mc admin info minio

echo "PROCESS > creating buckets..."
for bucket in "${buckets[@]}"
do
  mc mb minio/${bucket}
  sleep 1
done
echo "SUCCESS > buckets created."

echo "DISPLAY > minio buckets tree"
mc tree minio

echo "FINISH > keeping container up & running by laucnhing shell"
sh

velero-minio/resources/velero/scripts/auto-annotate-v1.1.0b.py
__version__ = '1.1.0b'
#!/usr/local/bin/python3
from kubernetes import client, config
from openshift.dynamic import DynamicClient
import sys
import urllib3
from tabulate import tabulate
import argparse

##########################################################################
################################ VARIABLES ###############################
##########################################################################

data = []

##########################################################################
##########################################################################
##########################################################################

#config.load_kube_config()

def parse():
    parser = argparse.ArgumentParser(description='Auto annotate your pods for Velero Restic PVC backup. Annotation is done at Deployment level.')
    parser.add_argument('namespace', metavar='namespace-to-tag', type=str, help='the namespace you want to inspect for velero backup tagging')
    parser.add_argument('--version', action='version', version='%(prog)s '+__version__)
    args = parser.parse_args()
    return args

def setup_config():
    k8s_client = config.new_client_from_config()
    dyn_client = DynamicClient(k8s_client)
    return dyn_client

def load_table(namespace, dyn_client):
    #dyn_client = setup_config()

    ocp_pod = dyn_client.resources.get(api_version='v1', kind='Pod')
    pods = ocp_pod.get(namespace=namespace)

    for pod in pods.items:
        for volume in pod.spec.volumes:
            if volume.get('persistentVolumeClaim'):
                volName = volume.get('name')
                pvcName = volume.get('persistentVolumeClaim').get('claimName')
                print(pod.metadata.name+" has a volume "+volName+" with PVC "+pvcName+" to backup")
                for controllerInfo in pod.metadata.ownerReferences:
                    if controllerInfo.get('kind') == "StatefulSet":
                        print("we have a statefulset")
                        fill_data(namespace,pod.metadata.name,volName,pvcName,controllerInfo.get('name'),controllerInfo.get('kind'),controllerInfo.get('name'),controllerInfo.get('kind'))
                    else:
                        ocp_controller = dyn_client.resources.get(api_version='v1', kind=controllerInfo.get('kind'))
                        controller = ocp_controller.get(name=controllerInfo.get('name'),namespace=namespace)
                        for deployerInfo in controller.metadata.ownerReferences:
                            fill_data(namespace,pod.metadata.name,volName,pvcName,controllerInfo.get('name'),controllerInfo.get('kind'),deployerInfo.get('name'),deployerInfo.get('kind'))
    print(tabulate(data, headers=['Namespace', 'Pod', 'Volume-name', 'PVC', 'Controller', 'Controller-kind', 'Deployer', 'Deploy-kind'], tablefmt="fancy_grid"))

def fill_data(namespace, podName, volName, pvcName, controllerName, controllerKind, deployerName, deployerKind):
    if not data:
        data.append([namespace, podName, volName, pvcName, controllerName, controllerKind, deployerName, deployerKind])
    else:
        i = 0
        for row in data:
            i += 1
            if row[1] == podName:
                row[2] = row[2] +","+ volName
                break
            elif i == len(data):
                data.append([namespace, podName, volName, pvcName, controllerName, controllerKind, deployerName, deployerKind])
                break

def proceed_tag(dyn_client):
    confirm = input('proceed ? [Y/n]:')
    if confirm != 'Y':
        exit(0)

    for row in data:
        namespace = row[0]
        name = row[6]
        kind = row[7]
        volumes = row[2]

        body = {
            'kind': kind,
            'apiVersion': 'v1',
            'metadata': {'name': name},
            'spec': {
                'template': {'metadata': {'annotations': {'backup.velero.io/backup-volumes': volumes}}},
            }
        }
        print(body)
        deploy = dyn_client.resources.get(api_version='v1', kind=kind)
        deploy.patch(body=body, namespace=namespace)


def main():
    #Disabling warning messages
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
    #Parsing terminal arguments
    args = parse()
    #Execute program
    dyn_client = setup_config()
    load_table(args.namespace, dyn_client)
    proceed_tag(dyn_client)

if __name__ == "__main__":
    main()


toolchain-private-worker/old-versions/install-pw-ocp-offline-v1.0.0.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Owned
__author__ = "IBM"
__copyright__ = "Copyright 2020, IBM GTS"
__credits__ = ["Frederic Spiers", "-","-"]
__license__ = "----"
__version__ = "1.0.0"
__date__ = "03/07/2020"
__maintainers__ = ["Frederic Spiers"]
__email__ = "frederic.spiers-isc.france@ibm.com"
__status__ = "Development"

############################################################################
################################## Import ##################################
############################################################################

import os
import sys
import requests
import fileinput
import re
import yaml
from timeit import default_timer as timer
from datetime import timedelta

############################################################################
################################# Variables ################################
############################################################################

#REGISTRY
region="eu-de"
OCP_private_registry="image-registry.openshift-image-registry.svc:5000"
dockerio_mapping_prefix=""
install_filename="updated-private-worker-install-ocp.yaml"

#PROXY
HTTPS_PROXY='http://dev6:dev6@10.194.216.124:3128'
HTTP_PROXY=''

#INT1
int1 =	{
    "OCP_API": "api.ocp4-int1.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int1.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "fQB6g-CpNPP-mEKw5-9ABXE"
}
#INT2
int2 =	{
    "OCP_API": "api.ocp4-int2.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int2.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": ""
}
#DEV5
dev5 =	{
    "OCP_API": "api.ocp4-dev5.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-dev5.devibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "tbbxZ-sqE7P-hNqhx-T5TDt"
}

#DEV6
dev6 =	{
    "OCP_API": "api.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.devibm.local",
    "OCP_user": "",
    "OCP_pwd": ""
}

#REGEX
patterns_list = {
    'gcr.io' : OCP_private_registry,
    'ibmcom' : OCP_private_registry+'/ibmcom',
    '@sha.*' : '',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
}

gcr_img = re.compile('^([\\s\\t]*gcr.io\\/tekton-releases\\/.*\\:.*@sha.*)$')
ibmcom_img = re.compile('^([\\s\\t]*.*ibmcom\\/.*\\:.*)$')
invisible_char = '[\\s\\t\\n]+'
digest_char = '@sha.*'

############################################################################
################################# Functions ################################
############################################################################

def get_pw_yaml():
    #os.system('curl -o '+install_filename+' https://private-worker-service.'+region+'.devops.cloud.ibm.com/install')
    r = requests.get('https://private-worker-service.'+region+'.devops.cloud.ibm.com/install', allow_redirects=True)
    open(install_filename, 'wb').write(r.content)

def select_img_name(string):
    tab = string.split('/')
    return tab[-1]

def replace_list_pattern(filename, pattern_list):
    file_to_modify = open(filename, 'rt')
    data = file_to_modify.read()
    for k, v in pattern_list.items():
        data = re.sub(k, v, data)
    file_to_modify.close()
    file_to_modify = open(filename, 'wt')
    file_to_modify.write(data)
    file_to_modify.close()

def configure_pw_access(filename):
    ro_data = open(filename, 'rt')
    ymls = list(yaml.safe_load_all(ro_data))
    for yml in ymls:
        if yml['kind'] == 'Deployment' and yml['metadata']['name'] == 'private-worker-agent':
            ymlEnv = (yml['spec']['template']['spec']['containers'][0]['env'])
            ymlEnv.append({'name': 'HTTPS_PROXY', 'value': HTTPS_PROXY})
            ymlEnv.append({'name': 'NO_PROXY', 'value': '172.30.0.1'})  
    ro_data.close()
    rw_data = open(filename, 'wt')
    yaml.dump_all(ymls, rw_data)
    rw_data.close()

def check_regex(string, pattern):
    if pattern.search(string):
        return True
    else:
        pass

def get_original_img():
    img_to_download = []
    data = open(install_filename, 'r')
    for line in data:
        if 'gcr.io' in line or 'ibmcom' in line:
            if check_regex(line, gcr_img):
                line = re.sub(invisible_char, '', re.sub(digest_char, '', line))
                img_to_download.append(line)
            elif check_regex(line, ibmcom_img):
                line = re.sub(invisible_char, '', re.sub(digest_char, '', line))
                line = re.sub("image:", '', line)
                img_to_download.append(line)
            else:
                pass
    return img_to_download

def fomat_img_list(image_list, pattern_list):
    formated_list = []
    for img in image_list:
        formated_img = img
        for k, v in pattern_list.items():
            formated_img = re.sub(k, v, formated_img)
        formated_list.append(formated_img)
    return formated_list

def OCP_Login(creds):
    print('OC login : https://'+creds.get("OCP_API"))
    os.system('/bin/oc43.sh login --insecure-skip-tls-verify=true https://'+creds.get("OCP_API")+' -u '+creds.get("OCP_user")+' -p '+creds.get("OCP_pwd"))
    token = os.popen('echo $(/bin/oc43.sh whoami -t)').read()
    token = re.sub(invisible_char, '', token)
    return token

def download_locally(img_list):
    for img in img_list:
        print('[ SKOPEO PULL ] > COPYING IMAGE FROM [[ docker://'+img+' ]] to [[ dir:/tmp/'+select_img_name(img)+' ]]')
        os.system('skopeo copy '+'docker://'+img+' '+'dir:/tmp/'+select_img_name(img))

def upload_OCP(img_list, target, token):
    for img in img_list:
        print('[ SKOPEO PUSH ] > COPYING IMAGE FROM [[ dir:/tmp/'+select_img_name(img)+' ]] to [[ docker://'+img)
        os.system('skopeo copy '+'dir:/tmp/'+select_img_name(img)+' '+'docker://'+img+' '+'--dest-creds '+target.get("OCP_user")+':'+token+' '+'--dest-tls-verify=false')

def install_pw():
    print('[ OCP ] > INSTALLING PRIVATE WORKER')
    os.system('/bin/oc43.sh apply -f '+install_filename)
    print('[ OCP ] > APPLYING PULL IMAGE POLICY')
    os.system('/bin/oc43.sh policy add-role-to-user system:image-puller system:serviceaccount:tekton-pipelines:tekton-pipelines-controller --namespace=tekton-releases')
    os.system('/bin/oc43.sh policy add-role-to-user system:image-puller system:serviceaccount:tekton-pipelines:private-worker-agent --namespace=tekton-releases')
    print('[ OCP ] > APPLYING SERVICE ACCOUNT POLICY')
    os.system('/bin/oc43.sh adm policy add-scc-to-user anyuid system:serviceaccount:tekton-pipelines:tekton-pipelines-controller')

    #oc get workeragent



############################################################################
################################### Main ###################################
############################################################################

def main():
    # Timer
    start = timer()

    # Set variable
    env = dev6

    env_patterns_list = {
    'gcr.io' : env.get("OCP_registry"),
    'ibmcom' : env.get("OCP_registry")+'/ibmcom',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
    }

    # Program
    get_pw_yaml()
    original_list = get_original_img()
    download_locally(original_list)
    formated_list = fomat_img_list(original_list, env_patterns_list)
    token = OCP_Login(env)
    upload_OCP(formated_list, env, token)
    replace_list_pattern(install_filename, patterns_list)
    configure_pw_access(install_filename)
    install_pw()

    # Timer
    end = timer()
    print('[[ Elapsed time : '+str(timedelta(seconds=end-start))+' ]]')
    
if __name__ == "__main__":
    main()


toolchain-private-worker/old-versions/install-pw-ocp-offline-v1.0.0b.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Owned
__author__ = "IBM"
__copyright__ = "Copyright 2020, IBM GTS"
__credits__ = ["Frederic Spiers", "-","-"]
__license__ = "----"
__version__ = "1.0.0b"
__date__ = "25/06/2020"
__maintainers__ = ["Frederic Spiers"]
__email__ = "frederic.spiers-isc.france@ibm.com"
__status__ = "Development"

############################################################################
################################## Import ##################################
############################################################################

import os
import sys
import requests
import fileinput
import re
from timeit import default_timer as timer
from datetime import timedelta

############################################################################
################################# Variables ################################
############################################################################

#REGISTRY
region="eu-de"
OCP_private_registry="image-registry.openshift-image-registry.svc:5000"
dockerio_mapping_prefix=""
install_filename="updated-private-worker-install-ocp.yaml"

#INT1
int1 =	{
    "OCP_API": "api.ocp4-int1.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int1.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "fQB6g-CpNPP-mEKw5-9ABXE"
}
#INT2
int2 =	{
    "OCP_API": "api.ocp4-int2.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int2.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": ""
}
#DEV5
dev5 =	{
    "OCP_API": "api.ocp4-dev5.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-dev5.devibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "tbbxZ-sqE7P-hNqhx-T5TDt"
}

#DEV6
dev6 =	{
    "OCP_API": "api.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.devibm.local",
    "OCP_user": "",
    "OCP_pwd": ""
}

#REGEX
patterns_list = {
    'gcr.io' : OCP_private_registry,
    'ibmcom' : OCP_private_registry+'/ibmcom',
    '@sha.*' : '',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
}

gcr_img = re.compile('^([\\s\\t]*gcr.io\\/tekton-releases\\/.*\\:.*@sha.*)$')
ibmcom_img = re.compile('^([\\s\\t]*.*ibmcom\\/.*\\:.*)$')
invisible_char = '[\\s\\t\\n]+'
digest_char = '@sha.*'

############################################################################
################################# Functions ################################
############################################################################

def get_pw_yaml():
    #os.system('curl -o '+install_filename+' https://private-worker-service.'+region+'.devops.cloud.ibm.com/install')
    r = requests.get('https://private-worker-service.'+region+'.devops.cloud.ibm.com/install', allow_redirects=True)
    open(install_filename, 'wb').write(r.content)

def select_img_name(string):
    tab = string.split('/')
    return tab[-1]

def replace_list_pattern(filename, pattern_list):
    file_to_modify = open(filename, 'rt')
    data = file_to_modify.read()
    for k, v in pattern_list.items():
        data = re.sub(k, v, data)
    file_to_modify.close()
    file_to_modify = open(filename, 'wt')
    file_to_modify.write(data)
    file_to_modify.close()

def check_regex(string, pattern):
    if pattern.search(string):
        return True
    else:
        print("doesn't seem to be a valid entry for : "+string)

def get_original_img():
    img_to_download = []
    data = open(install_filename, 'r')
    for line in data:
        if 'gcr.io' in line or 'ibmcom' in line:
            if check_regex(line, gcr_img):
                line = re.sub(invisible_char, '', line)
                img_to_download.append(line)
            elif check_regex(line, ibmcom_img):
                line = re.sub(invisible_char, '', line)
                line = re.sub("image:", '', line)
                img_to_download.append(line)
            else:
                pass
    return img_to_download

def fomat_img_list(image_list, pattern_list):
    cleaned_list = []
    formated_list = []
    for img in image_list:
        cleaned_img = re.sub(digest_char, '', img)
        cleaned_list.append(cleaned_img)
        formated_img = cleaned_img
        for k, v in pattern_list.items():
            formated_img = re.sub(k, v, formated_img)
        formated_list.append(formated_img)
    return cleaned_list, formated_list

def OCP_Login(creds):
    print('OC login : https://'+creds.get("OCP_API"))
    os.system('/bin/oc43.sh login --insecure-skip-tls-verify=true https://'+creds.get("OCP_API")+' -u '+creds.get("OCP_user")+' -p '+creds.get("OCP_pwd"))
    token = os.popen('echo $(/bin/oc43.sh whoami -t)').read()
    token = re.sub(invisible_char, '', token)
    return token

def download_locally(img_list):
    for img in img_list:
        short_img = select_img_name(img)

        print('[ SKOPEO PULL ] > COPYING IMAGE FROM [[ docker://'+img+' ]] to [[ dir:/tmp/'+short_img+' ]]')
        os.system('skopeo copy '+'docker://'+img+' '+'dir:/tmp/'+short_img)

def upload_OCP(img_list, target):
    for img in img_list:
        short_img = select_img_name(re.sub(digest_char, '', img))
        token = OCP_Login(target)

        print('[ SKOPEO PUSH ] > COPYING IMAGE FROM [[ dir:/tmp/'+short_img+' ]] to [[ docker://'+img)
        os.system('skopeo copy '+'dir:/tmp/'+short_img+' '+'docker://'+img+' '+'--dest-creds '+target.get("OCP_user")+':'+token+' '+'--dest-tls-verify=false')


def display_downloaded_img():
    print('[[ List of locally dowloaded images ]]\n')
    os.system('ls -1 /tmp')

def print_original_img():
    data = open(install_filename, 'r')
    for line in data:
        if 'gcr.io' in line or 'ibmcom' in line:
            if not '>-' in line:
                if check_regex(line, gcr_img) or check_regex(line, ibmcom_img):
                    print('original image : '+line)

def print_result():
    data = open(install_filename, 'r')
    for line in data:
        if OCP_private_registry in line:
            if not '>-' in line:
                print('new image name : '+line)

def print_list(img_list):
    print('\n NEW LIST :')
    for img in img_list:
        print(img)

############################################################################
################################### Main ###################################
############################################################################

def main():
    # Timer
    start = timer()

    # Set variable
    env = dev6

    env_patterns_list = {
    'gcr.io' : env.get("OCP_registry"),
    'ibmcom' : env.get("OCP_registry")+'/ibmcom',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
    }

    # Program
    get_pw_yaml()
    print_original_img()
    original_list = get_original_img()

    cleaned_list, formated_list = fomat_img_list(original_list, env_patterns_list)

    download_locally(cleaned_list)
    display_downloaded_img()
    upload_OCP(formated_list, env)
    replace_list_pattern(install_filename, patterns_list)
    #print_result()

    # Timer
    end = timer()
    print('[[ Elaspsed time : '+str(timedelta(seconds=end-start))+' ]]')
    
if __name__ == "__main__":
    main()

toolchain-private-worker/install-pw-ocp-offline.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Owned
__author__ = "IBM"
__copyright__ = "Copyright 2020, IBM GTS"
__credits__ = ["Frederic Spiers", "-","-"]
__license__ = "----"
__version__ = "2.0.0"
__date__ = "04/07/2020"
__maintainers__ = ["Frederic Spiers"]
__email__ = "frederic.spiers-isc.france@ibm.com"
__status__ = "Development"

############################################################################
################################## Import ##################################
############################################################################

import os
import sys
import requests
import fileinput
import re
import yaml
from timeit import default_timer as timer
from datetime import timedelta

############################################################################
################################# Variables ################################
############################################################################

#REGISTRY
region="eu-de"
OCP_private_registry="image-registry.openshift-image-registry.svc:5000"
dockerio_mapping_prefix=""
install_filename="updated-private-worker-install-ocp.yaml"

#INT1
int1 =	{
    "OCP_API": "api.ocp4-int1.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int1.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "fQB6g-CpNPP-mEKw5-9ABXE"
}
#INT2
int2 =	{
    "OCP_API": "api.ocp4-int2.intibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-int2.intibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": ""
}
#DEV5
dev5 =	{
    "OCP_API": "api.ocp4-dev5.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.ocp4-dev5.devibm.local",
    "OCP_user": "kubeadmin",
    "OCP_pwd": "tbbxZ-sqE7P-hNqhx-T5TDt"
}

#DEV6
dev6 =	{
    "OCP_API": "api.devibm.local:6443",
    "OCP_registry": "default-route-openshift-image-registry.apps.devibm.local",
    "OCP_user": "",
    "OCP_pwd": ""
}

#REGEX
patterns_list = {
    'gcr.io' : OCP_private_registry,
    'ibmcom' : OCP_private_registry+'/ibmcom',
    '@sha.*' : '',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
}

gcr_img = re.compile('^([\\s\\t]*gcr.io\\/tekton-releases\\/.*\\:.*@sha.*)$')
ibmcom_img = re.compile('^([\\s\\t]*.*ibmcom\\/.*\\:.*)$')
invisible_char = '[\\s\\t\\n]+'
digest_char = '@sha.*'

############################################################################
################################# Functions ################################
############################################################################

def setEnv():
    if os.environ.get('TARGET') is None:
        print("Error : please set ENV VAR 'TARGET' on your system" )
        exit(1)
    else:
        if os.environ.get('TARGET') in globals():
            env = globals()[os.environ.get('TARGET')]
            return env
        else:
            print('Error : Environment variable '+os.environ.get('TARGET')+' does not exist. Please choose a proper one.')
            exit(1)

def setCreds(env):
    if os.environ.get('OC_USER') is None and os.environ.get('OC_PWD') is None:
        pass
    else:
        env['OCP_user'] = os.environ.get('OC_USER')
        env['OCP_pwd'] = os.environ.get('OC_PWD')

def get_pw_yaml():
    #os.system('curl -o '+install_filename+' https://private-worker-service.'+region+'.devops.cloud.ibm.com/install')
    r = requests.get('https://private-worker-service.'+region+'.devops.cloud.ibm.com/install', allow_redirects=True)
    open(install_filename, 'wb').write(r.content)

def select_img_name(string):
    tab = string.split('/')
    return tab[-1]

def replace_list_pattern(filename, pattern_list):
    file_to_modify = open(filename, 'rt')
    data = file_to_modify.read()
    for k, v in pattern_list.items():
        data = re.sub(k, v, data)
    file_to_modify.close()
    file_to_modify = open(filename, 'wt')
    file_to_modify.write(data)
    file_to_modify.close()

def configure_pw_proxy(filename):
    ro_data = open(filename, 'rt')
    ymls = list(yaml.safe_load_all(ro_data))
    for yml in ymls:
        if yml['kind'] == 'Deployment' and yml['metadata']['name'] == 'private-worker-agent':
            ymlEnv = (yml['spec']['template']['spec']['containers'][0]['env'])
            ymlEnv.append({'name': 'HTTPS_PROXY', 'value': os.environ.get('OC_HTTPS_PROXY')})
            ymlEnv.append({'name': 'NO_PROXY', 'value': '172.30.0.1'})
    ro_data.close()
    rw_data = open(filename, 'wt')
    yaml.dump_all(ymls, rw_data)
    rw_data.close()

def check_regex(string, pattern):
    if pattern.search(string):
        return True
    else:
        pass

def get_original_img():
    img_to_download = []
    data = open(install_filename, 'r')
    for line in data:
        if 'gcr.io' in line or 'ibmcom' in line:
            if check_regex(line, gcr_img):
                line = re.sub(invisible_char, '', re.sub(digest_char, '', line))
                img_to_download.append(line)
            elif check_regex(line, ibmcom_img):
                line = re.sub(invisible_char, '', re.sub(digest_char, '', line))
                line = re.sub("image:", '', line)
                img_to_download.append(line)
            else:
                pass
    return img_to_download

def fomat_img_list(image_list, pattern_list):
    formated_list = []
    for img in image_list:
        formated_img = img
        for k, v in pattern_list.items():
            formated_img = re.sub(k, v, formated_img)
        formated_list.append(formated_img)
    return formated_list

def OCP_Login(creds):
    print('OC login : https://'+creds.get("OCP_API"))
    os.system('/bin/oc43.sh login --insecure-skip-tls-verify=true https://'+creds.get("OCP_API")+' -u '+creds.get("OCP_user")+' -p '+creds.get("OCP_pwd"))
    token = os.popen('echo $(/bin/oc43.sh whoami -t)').read()
    token = re.sub(invisible_char, '', token)
    return token

def download_locally(img_list):
    for img in img_list:
        print('[ SKOPEO PULL ] > COPYING IMAGE FROM [[ docker://'+img+' ]] to [[ dir:/tmp/'+select_img_name(img)+' ]]')
        os.system('skopeo copy '+'docker://'+img+' '+'dir:/tmp/'+select_img_name(img))

def upload_OCP(img_list, target, token):
    for img in img_list:
        print('[ SKOPEO PUSH ] > COPYING IMAGE FROM [[ dir:/tmp/'+select_img_name(img)+' ]] to [[ docker://'+img)
        os.system('skopeo copy '+'dir:/tmp/'+select_img_name(img)+' '+'docker://'+img+' '+'--dest-creds '+target.get("OCP_user")+':'+token+' '+'--dest-tls-verify=false')

def install_pw():
    print('[ OCP ] > INSTALLING PRIVATE WORKER')
    os.system('/bin/oc43.sh apply -f '+install_filename)
    print('[ OCP ] > APPLYING PULL IMAGE POLICY')
    os.system('/bin/oc43.sh policy add-role-to-user system:image-puller system:serviceaccount:tekton-pipelines:tekton-pipelines-controller --namespace=tekton-releases')
    os.system('/bin/oc43.sh policy add-role-to-user system:image-puller system:serviceaccount:tekton-pipelines:private-worker-agent --namespace=tekton-releases')
    print('[ OCP ] > APPLYING SERVICE ACCOUNT POLICY')
    os.system('/bin/oc43.sh adm policy add-scc-to-user anyuid system:serviceaccount:tekton-pipelines:tekton-pipelines-controller')

############################################################################
################################### Main ###################################
############################################################################

def main():
    # Timer
    start = timer()

    # Set variables
    env = setEnv()
    setCreds(env)

    env_patterns_list = {
    'gcr.io' : env.get("OCP_registry"),
    'ibmcom' : env.get("OCP_registry")+'/ibmcom',
    '\\/tekton-releases\\/.*\\/' : '/tekton-releases/'
    }

    # Program
    get_pw_yaml()
    original_list = get_original_img()
    download_locally(original_list)
    formated_list = fomat_img_list(original_list, env_patterns_list)
    token = OCP_Login(env)
    upload_OCP(formated_list, env, token)
    replace_list_pattern(install_filename, patterns_list)
    if os.environ.get('OC_HTTPS_PROXY'):
        configure_pw_proxy(install_filename)
    install_pw()

    # Timer
    end = timer()
    print('[[ Elapsed time : '+str(timedelta(seconds=end-start))+' ]]')
    
if __name__ == "__main__":
    main()